--- UICheckBoxButton.lua	2015-07-14 11:14:26.000000000 +0800
+++ UICheckBoxButton.lua-new	2015-07-14 11:13:58.000000000 +0800
@@ -97,22 +97,6 @@
         or state == UICheckBoxButton.ON_DISABLED,
         string.format("UICheckBoxButton:setButtonImage() - invalid state %s", tostring(state)))
     UICheckBoxButton.super.setButtonImage(self, state, image, ignoreEmpty)
-    if state == UICheckBoxButton.OFF then
-        if not self.images_[UICheckBoxButton.OFF_PRESSED] then
-            self.images_[UICheckBoxButton.OFF_PRESSED] = image
-        end
-        if not self.images_[UICheckBoxButton.OFF_DISABLED] then
-            self.images_[UICheckBoxButton.OFF_DISABLED] = image
-        end
-    elseif state == UICheckBoxButton.ON then
-        if not self.images_[UICheckBoxButton.ON_PRESSED] then
-            self.images_[UICheckBoxButton.ON_PRESSED] = image
-        end
-        if not self.images_[UICheckBoxButton.ON_DISABLED] then
-            self.images_[UICheckBoxButton.ON_DISABLED] = image
-        end
-    end
-
     return self
 end
 
