--- UIPushButton.lua	2015-07-14 11:11:46.000000000 +0800
+++ UIPushButton.lua-new	2015-07-14 11:10:11.000000000 +0800
@@ -78,16 +78,6 @@
         or state == UIPushButton.DISABLED,
         string.format("UIPushButton:setButtonImage() - invalid state %s", tostring(state)))
     UIPushButton.super.setButtonImage(self, state, image, ignoreEmpty)
-
-    if state == UIPushButton.NORMAL then
-        if not self.images_[UIPushButton.PRESSED] then
-            self.images_[UIPushButton.PRESSED] = image
-        end
-        if not self.images_[UIPushButton.DISABLED] then
-            self.images_[UIPushButton.DISABLED] = image
-        end
-    end
-
     return self
 end
 
